app/services/api/teachers/withdraw.rb

```diff
@@ -0,0 +1,106 @@
+module API::Teachers
+  class Withdraw
+    include ActiveModel::Model
+    include ActiveModel::Attributes
+
+    COURSE_IDENTIFIERS = %w[ecf-mentor ecf-induction].freeze
+
+    attribute :lead_provider_id
+    attribute :teacher_api_id
+    attribute :reason
+    attribute :course_identifier
+
+    validates :lead_provider_id, presence: { message: "Enter a '#/lead_provider_id'." }
+    validates :teacher_api_id, presence: { message: "Enter a '#/teacher_api_id'." }
+    validates :course_identifier, presence: { message: "Enter a '#/course_identifier'." }
+    validates :reason, presence: { message: "Enter a '#/reason'." }
+
+    validate :lead_provider_exists
+    validate :teacher_training_exists
+    validates :course_identifier, inclusion: {
+      in: COURSE_IDENTIFIERS,
+      message: "The entered '#/course_identifier' is not recognised for the given participant. Check details and try again."
+    }, allow_blank: true
+    validates :reason, inclusion: {
+      in: TrainingPeriod.withdrawal_reasons.values.map(&:dasherize),
+      message: "The entered '#/reason' is not recognised for the given participant. Check details and try again."
+    }, allow_blank: true
+    validate :not_already_withdrawn
+
+    def withdraw
+      return false unless valid?
+
+      ActiveRecord::Base.transaction do
+        training_period.withdrawn_at = Time.zone.now
+        training_period.withdrawal_reason = reason.underscore
+        training_period.finished_on = [training_period.finished_on, Time.zone.today].compact.min
```

Comment:
Can we use `training_period.withdrawn_at` here instead of `Time.zone.today`? I'm just thinking a lot of LP stuff happens on the hour and that would avoid potential edge-cases where the `finished_on` could be after the `withdrawn_at` if they do something at midnight